[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33464 [TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass.**

I added a python-exposed knob to register this pass in custom passes pipeline. If the knob is not used, the pass is not registered and thus not run at all.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

Commit:
[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass.

The pass registers itself as a custom backend pass and runs after the existing CUDA fuser. In order to actually see the new pass in action (when it's done) one would need to disable the existing fuser with `torch._C._jit_override_can_fuse_on_gpu(false)`.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass."


The pass registers itself as a custom backend pass and runs after the existing CUDA fuser. In order to actually see the new pass in action (when it's done) one would need to disable the existing fuser with `torch._C._jit_override_can_fuse_on_gpu(false)`.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass."


I added a python-exposed knob to register this pass in custom passes pipeline. If the knob is not used, the pass is not registered and thus not run at all.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass."


I added a python-exposed knob to register this pass in custom passes pipeline. If the knob is not used, the pass is not registered and thus not run at all.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/passes/tensorexpr_fuser.h:
```diff
@@ -0,0 +1,18 @@
+struct Graph;
+
+// Run TensorExpressions-based fuser.
+TORCH_API void fuseTensorExprs(std::shared_ptr<Graph>& graph);
```
Do we need to expose this in the header?  Not that it's a big deal.

ZolotukhinM replied:
I think it's good to have it exposed at least for testing purposes.

bertmaher reviewed (approved):
lgtm!

suo commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -0,0 +1,205 @@
+  // Grab the iterator below consumer.  We'll use that to determine
+  // where to resume iteration, even if consumer gets relocated within
+  // the block.
+  auto iter = --consumer->reverseIterator();
```
I find the iterator flow here a little difficult to understand—decrementing a reverse iterator, double incrementing an iterator, doing it prefix, etc. Is there some way we can use the `next()`/`prev()` APIs to make this a little clearer?

ZolotukhinM replied:
There should be an easier way, yes - e.g. `create_autodiff_subgraph` does a similar thing and doesn't have this decrements/double-increments. I will follow up on this as we start adding actual code.

PS: I also hope this all becomes much simpler once we have functional subgraphs.

suo reviewed (approved):
lgtm as a skeleton. We may need to revise post-functionalization pass (cc @eellison) but that should not block this

Commit:
Rebase on "[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass."


I added a python-exposed knob to register this pass in custom passes pipeline. If the knob is not used, the pass is not registered and thus not run at all.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

[ghstack-poisoned]
